File: app/src/main/java/org/neidhardt/dynamicsoundboard/dialog/SoundSettingsDialog.java
package org.neidhardt.dynamicsoundboard.dialog;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.FragmentManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.customview.edittext.CustomEditText;
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData;
import org.neidhardt.dynamicsoundboard.dao.SoundSheet;
import org.neidhardt.dynamicsoundboard.dialog.addnewsoundfromintent.CustomSpinner;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.soundcontrol.SoundSheetFragment;
import org.neidhardt.dynamicsoundboard.soundmanagement.ServiceManagerFragment;
import org.neidhardt.dynamicsoundboard.soundsheet.SoundSheetsManagerFragment;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;

/**
 * Created by eric.neidhardt on 23.02.2015.
 */
public class SoundSettingsDialog extends BaseDialog implements View.OnClickListener, CompoundButton.OnCheckedChangeListener
{
	private static final String TAG = SoundSettingsDialog.class.getName();

	private static final String KEY_PLAYER_ID = "org.neidhardt.dynamicsoundboard.dialog.SoundSettingsDialog.playerId";
	private static final String KEY_FRAGMENT_TAG = "org.neidhardt.dynamicsoundboard.dialog.SoundSettingsDialog.fragmentTag";

	private CustomEditText soundName;
	private CustomEditText soundSheetName;
	private CustomSpinner soundSheetSpinner;
	private CheckBox addNewSoundSheet;

	private String playerId;
	private String fragmentTag;
	private int indexOfCurrentFragment = -1;
	private EnhancedMediaPlayer player;

	public static void showInstance(FragmentManager manager, MediaPlayerData playerData)
	{
		SoundSettingsDialog dialog = new SoundSettingsDialog();

		Bundle args = new Bundle();
		args.putString(KEY_PLAYER_ID, playerData.getPlayerId());
		args.putString(KEY_FRAGMENT_TAG, playerData.getFragmentTag());

		dialog.setArguments(args);

		dialog.show(manager, TAG);
	}

	@Override
	public void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		Bundle args = this.getArguments();
		if (args != null)
		{
			this.playerId = args.getString(KEY_PLAYER_ID);
			this.fragmentTag = args.getString(KEY_FRAGMENT_TAG);
			this.player = this.getServiceManagerFragment().getSoundService().searchForId(this.fragmentTag, this.playerId);
		}
	}

	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		@SuppressLint("InflateParams") View view = this.getActivity().getLayoutInflater().inflate(R.layout.dialog_sound_settings_layout, null);

		this.soundName = (CustomEditText)view.findViewById(R.id.et_name_file);
		this.soundSheetName = (CustomEditText)view.findViewById(R.id.et_name_new_sound_sheet);
		this.soundSheetSpinner = (CustomSpinner)view.findViewById(R.id.s_sound_sheets);
		this.addNewSoundSheet = (CheckBox)view.findViewById(R.id.cb_add_new_sound_sheet);

		this.addNewSoundSheet.setOnCheckedChangeListener(this);

		view.findViewById(R.id.b_cancel).setOnClickListener(this);
		view.findViewById(R.id.b_save).setOnClickListener(this);

		this.soundName.setText(this.player.getMediaPlayerData().getLabel());

		this.setAvailableSoundSheets();

		this.soundSheetName.setText(this.getSoundSheetManagerFragment().getSuggestedSoundSheetName());
		this.soundSheetName.setVisibility(View.GONE);

		AlertDialog.Builder builder = new AlertDialog.Builder(this.getActivity());
		builder.setView(view);

		return builder.create();
	}

	private void setAvailableSoundSheets()
	{
		List<SoundSheet> soundSheets = this.getSoundSheetManagerFragment().getAll();
		ArrayList<String> labels = new ArrayList<>();
		for (int i = 0; i < soundSheets.size(); i++)
		{
			if (soundSheets.get(i).getFragmentTag().equals(this.fragmentTag))
				this.indexOfCurrentFragment = i;
			labels.add(soundSheets.get(i).getLabel());
		}
		if (this.indexOfCurrentFragment == -1)
			throw new IllegalStateException(TAG + " Current fragment of sound " + this.player.getMediaPlayerData() + " is not found in list of sound sheets " + soundSheets);

		this.soundSheetSpinner.setItems(labels);
		this.soundSheetSpinner.setSelectedItem(this.indexOfCurrentFragment);
	}

	@Override
	public void onCheckedChanged(CompoundButton buttonView, boolean isChecked)
	{
		if (isChecked)
		{
			this.soundSheetSpinner.setVisibility(View.GONE);
			this.soundSheetName.setVisibility(View.VISIBLE);
		}
		else
		{
			this.soundSheetName.setVisibility(View.GONE);
			this.soundSheetSpinner.setVisibility(View.VISIBLE);
		}
	}

	@Override
	public void onClick(View v)
	{
		switch (v.getId())
		{
			case R.id.b_cancel:
				this.dismiss();
				break;
			case R.id.b_save:
				this.deliverResult();
				this.dismiss();
				break;
		}
	}

	private void deliverResult()
	{
		String soundLabel = this.soundName.getDisplayedText();
		int indexOfSelectedSoundSheet = this.soundSheetSpinner.getSelectedItemPosition();
		boolean addNewSoundSheet = this.addNewSoundSheet.isChecked();
		boolean hasSoundSheetChanged = addNewSoundSheet || indexOfSelectedSoundSheet != this.indexOfCurrentFragment;

		SoundSheetFragment soundSheetFragment = this.getSoundSheetFragment(this.fragmentTag);
		if (!hasSoundSheetChanged)
		{
			this.player.getMediaPlayerData().setLabel(soundLabel);
			soundSheetFragment.notifyDataSetChanged();
		}
		else
		{
			ServiceManagerFragment serviceManagerFragment = this.getServiceManagerFragment();
			serviceManagerFragment.getSoundService().removeSounds(asList(this.player));
			soundSheetFragment.notifyDataSetChanged();

			SoundSheetsManagerFragment soundSheetsManagerFragment = this.getSoundSheetManagerFragment();
			Uri uri = Uri.parse(this.player.getMediaPlayerData().getUri());
			if (addNewSoundSheet)
				soundSheetsManagerFragment.addSoundToSoundSheet(uri, soundLabel, soundLabel, null);
			else
				soundSheetsManagerFragment.addSoundToSoundSheet(uri, soundLabel, null, soundSheetsManagerFragment.get(this.fragmentTag));
		}
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/dialog/soundmanagement/SoundSettingsDialog.kt
package org.neidhardt.dynamicsoundboard.dialog.soundmanagement

import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.CheckBox
import kotlinx.android.synthetic.main.dialog_sound_settings_layout.view.*
import org.neidhardt.android_utils.views.CustomEditText
import org.neidhardt.android_utils.views.SimpleSpinner
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.manager.SoundSheetManager
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.dynamicsoundboard.persistance.model.NewMediaPlayerData
import org.neidhardt.dynamicsoundboard.persistance.model.NewSoundSheet
import org.neidhardt.utils.letThis
import java.util.*
import kotlin.properties.Delegates

/**
 * File created by eric.neidhardt on 23.02.2015.
 */
class SoundSettingsDialog : SoundSettingsBaseDialog() {

	override var fragmentTag: String by Delegates.notNull<String>()
	override var player: MediaPlayerController by Delegates.notNull<MediaPlayerController>()
	override var soundSheet: NewSoundSheet? = null

	private var soundName: CustomEditText? = null
	private var soundSheetName: CustomEditText? = null
	private var soundSheetSpinner: SimpleSpinner? = null
	private var addNewSoundSheet: CheckBox? = null

	private var indexOfCurrentFragment = -1

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		@SuppressLint("InflateParams") val view = this.activity.layoutInflater.inflate(R.layout.dialog_sound_settings_layout, null)

		this.soundName = view.et_name_file.letThis {
			it.text = this.player.mediaPlayerData.label
		}
		this.soundSheetName = view.et_name_new_sound_sheet.letThis {
			it.text = this.soundSheetManager.suggestedName
			it.visibility = View.GONE
		}
		this.soundSheetSpinner = view.s_sound_sheets?.letThis {
			this.setAvailableSoundSheets(it)
		}
		this.addNewSoundSheet = view.cb_add_new_sound_sheet.letThis {
			it.setOnCheckedChangeListener { view,isChecked ->
				this.soundSheetSpinner?.visibility = if (isChecked) View.GONE else View.VISIBLE
				this.soundSheetName?.visibility = if (isChecked) View.VISIBLE else View.GONE
			}
		}

		return AlertDialog.Builder(context).apply {
			this.setTitle(R.string.dialog_sound_settings_title)
			this.setView(view)
			this.setPositiveButton(R.string.dialog_save, { dialogInterface, i ->
				val hasLabelChanged = player.mediaPlayerData.label != soundName!!.displayedText
				deliverResult()
				dismiss()
				if (hasLabelChanged)
					RenameSoundFileDialog.show(fragmentManager, player.mediaPlayerData)
			})
			this.setNegativeButton(R.string.dialog_cancel, { dialogInterface, i -> dismiss()})
		}.create()
	}

	private fun setAvailableSoundSheets(spinner: SimpleSpinner) {
		val soundSheets = this.soundSheetManager.soundSheets
		val labels = ArrayList<String>()
		for (i in soundSheets.indices) {
			if (soundSheets[i].fragmentTag == this.fragmentTag)
				this.indexOfCurrentFragment = i
			labels.add(soundSheets[i].label)
		}
		if (this.indexOfCurrentFragment == -1)
			throw IllegalStateException(TAG + " Current fragment of sound " + this.player.mediaPlayerData + " is not found in list of sound sheets " + soundSheets)

		spinner.items = labels
		spinner.setSelection(this.indexOfCurrentFragment)
	}

	private fun deliverResult() {
		val soundLabel = this.soundName!!.displayedText
		val indexOfSelectedSoundSheet = this.soundSheetSpinner!!.selectedItemPosition

		val addNewSoundSheet = this.addNewSoundSheet!!.isChecked
		val hasSoundSheetChanged = addNewSoundSheet || indexOfSelectedSoundSheet != this.indexOfCurrentFragment

		if (!hasSoundSheetChanged) {
			this.player.mediaPlayerData.label = soundLabel
			this.soundManager.notifyHasChanged(this.player)
			return
		}

		val newSoundSheet = if (addNewSoundSheet) {
			val newSoundSheetLabel = this.soundSheetName?.displayedText ?: ""
			val newFragmentTag = SoundSheetManager.getNewFragmentTagForLabel(newSoundSheetLabel)
			val newSoundSheet = NewSoundSheet().apply {
				this.fragmentTag = newFragmentTag
				this.label = newSoundSheetLabel
			}
			this.soundSheetManager.add(newSoundSheet)
			newSoundSheet
		}
		else null

		if (hasSoundSheetChanged) {
			// remove the sound from manager
			this.soundManager.remove(this.soundSheet!!, listOf(this.player))

			// the player is added to the newly created soundSheet
			if (addNewSoundSheet) {
				this.player.mediaPlayerData.fragmentTag = newSoundSheet?.fragmentTag
				this.soundManager.add(newSoundSheet!!, this.player.mediaPlayerData)
			}
			else { // the player is added to the selected soundSheet
				val selectedSoundSheet = this.soundSheetManager.soundSheets[indexOfSelectedSoundSheet]
				this.player.mediaPlayerData.fragmentTag = selectedSoundSheet.fragmentTag
				this.soundManager.add(selectedSoundSheet, this.player.mediaPlayerData)
			}
		}
	}

	companion object {
		private val TAG = SoundSettingsDialog::class.java.name

		fun showInstance(manager: FragmentManager, playerData: NewMediaPlayerData) {
			val dialog = SoundSettingsDialog()
			addArguments(dialog, playerData.playerId, playerData.fragmentTag)
			dialog.show(manager, TAG)
		}
	}

}
